include/libvfio-user.h

```diff
@@ -607,45 +603,28 @@ typedef struct {
      */
     int (*transition)(vfu_ctx_t *vfu_ctx, vfu_migr_state_t state);
     
+    /*
+     * Function that is called to read `count` bytes of migration data into
+     * `buf`. The function must return the amount of data read or -1 on error,
+     * setting errno.
+     * 
+     * If the function returns zero, this is interpreted to mean that there is
```

Comment:
I discussed this with John - we agreed that we shouldn't deviate from the normal semantics. See https://github.com/nutanix/libvfio-user/pull/746#discussion_r1289806033.

Reply:
What I meant is that it's not clearly documented.

Reply:
What I'm trying to say is that the callback can return less data than requested, since thre might not be as much migration data available, right? Shouldn't this be documented?

Reply:
I did update it, `include/libvfio-user.h:609` https://github.com/nutanix/libvfio-user/pull/746/commits/977211fd484204649410b288627a76644e36c3fe
```diff
-     * setting errno.
+     * setting errno. The function may return less data than requested.
```

Reply:
Not sure what happenend and it didn't show whan I was reviewing it. What does it mean returning less than requested? Does it mean that there are no more migration data to be read? As if the client had done a read of `ret` size (in which case we would have returned exactly `ret`) followed by another read where we would have returned 0?

Reply:
"less data than requested" := 0 < ret < count

Reply:
I get that but it's not that what I had in mind. I guess the impelementor can return `< count` if there's not enough migration data or simply because it chooses to do so.

Reply:
There could be any amount of data to read, right? It's up to the implementation to decide when it's done, and exactly how much it returns each time - it might have good reasons to return something less than the given buffer, but then have more to return. For example, across a page boundary.
That's why I was arguing this should have read() style semantics.

Reply:
and in particular I think this comment is clear enough IMO